datalad/runner/protocol.py

```diff
@@ -41,13 +41,34 @@ class GeneratorMixIn:
         for result in runner.run(...):
             # do something, for example write to stdin of the subprocess
 
-    """
-    def __init__(self):
-        self.result_queue = deque()
 
+    Example for creating a generator-version of the StdOutCapture protocol, that
+    will yield tuples of (fd, data):
+
+        class GeneratorStdOutCapture(StdOutCapture, GeneratorMixIn):
+            def pipe_data_received(self, fd, data):
+                self.send_result((fd, data))
+
+    """
     def send_result(self, result):
+        """ This method sends a result that will be yielded by the generator
+
+        If ThreadedRunner.run() is called with a protocol that is derived from
+        this mixin-class, it will return a generator. The elements that the
+        generator yields have to be sent to the generator. This is done via this
+        method. Any `result` that is given to `GeneratorMixIn.send_result()`
+        will eventually be yielded by the generator. The order in which they
+        are yielded is the order in which they are "given" to
+        `GeneratorMixIn.send_result()`.
+        """
         self.result_queue.append(result)
 
+    @property
+    def result_queue(self) -> deque:
+        if not hasattr(self, '_result_queue'):
```

Comment:
IMHO adding a comment for later us might save some time

```suggestion
        # lazy assignment avoids needing __init__ for the MixIn class
        if not hasattr(self, '_result_queue'):
```